### PR TITLE
[FIX] Enable the posibility of adding more products of a gifted product as new line in the order.

### PR DESCRIPTION
**Pull request by alvaro-domatix**
I think this is a posible solution for the question I posed [here](https://github.com/OCA/sale-promotion/issues/30).

This code allows to add as new lines in the order more units of a product that has been gifted for some promotion.

### CODE REVIEW

**Review by chienandalu — COMMENTED** [review]
The fix looks good, but a new module should be made:

- This module doesn't depend on `website_sale_coupon`, so this method isn't available.
- It will happen with `sale_coupon_multiplier_free_product` which uses the same kind of discount lines.

**Comment by alvaro-domatix** [comment]
Alright! I'll do it. Any module name suggestion? @chienandalu

**Comment by chienandalu** [comment]
@pedrobaeza uses to be more inspired with names than me :sweat_smile:

**Comment by pedrobaeza** [comment]
`website_sale_coupon_multi_gift` should be the obvious choose.

**Comment by chienandalu** [comment]
It's not only compatibility with `sale_coupon_multi_gift` but with `sale_coupon_multiplier_free_product` as well (or any other using that kind of rewads)

**Comment by pedrobaeza** [comment]
Uhm, `website_sale_coupon_reward_line_discriminant`? Anyways, shouldn't this be done at `website_sale_coupon` level and it's a bug?

**Comment by alvaro-domatix** [comment]
> Uhm, `website_sale_coupon_reward_line_discriminant`? Anyways, shouldn't this be done at `website_sale_coupon` level and it's a bug?

No because in website_sale_coupon when you gift one product of your shop, it adds one order line for the product (where you can change the quantity) and one for the gift. multi_gift module adds one unique line with 100%discount so you can't modify the quantity as it is a "reward_line".

I will use website_sale_coupon_reward_line_discriminant as the name if you both agree.

**Comment by chienandalu** [comment]
:+1:

File: website_sale_coupon_reward_line_discriminant/__manifest__.py [hunk 1]
```diff
@@ -0,0 +1,14 @@
...
+# License AGPL-3.0 or later (https://www.gnu.org/licenses/agpl).
+
+{
+    "name": "Website Sale Coupon Reward Line Discriminant",
+    "version": "13.0.1.0.0",
+    "author": "Domatix, Odoo Community Association (OCA)",
+    "website": "https://github.com/OCA/sale-promotion",
+    "license": "AGPL-3",
+    "category": "Sale",
+    "depends": ["website_sale"],
```
Comment (chienandalu) [inline website_sale_coupon_reward_line_discriminant/__manifest__.py]:
```suggestion
    "depends": ["website_sale_coupon"],
```

File: website_sale_coupon_reward_line_discriminant/models/sale_order.py [hunk 2]
```diff
@@ -0,0 +1,14 @@
+# Copyright 2021 Domatix - Álvaro López
+# License AGPL-3.0 or later (https://www.gnu.org/licenses/agpl).
+from odoo import models
+
+
+class SaleOrder(models.Model):
+    _inherit = "sale.order"
+
+    def _cart_find_product_line(self, product_id=None, line_id=None, **kwargs):
+        lines = super(SaleOrder, self)._cart_find_product_line(
```
Comment (chienandalu) [inline website_sale_coupon_reward_line_discriminant/models/sale_order.py]:
```suggestion
        lines = super()._cart_find_product_line(
```

File: website_sale_coupon_reward_line_discriminant/readme/DESCRIPTION.rst [hunk 3]
```diff
@@ -0,0 +1 @@
+This module allows to modify quantity of a sale order line that are a product gift reward adding a new line to manage not reward quantities.
```
Comment (chienandalu) [inline website_sale_coupon_reward_line_discriminant/readme/DESCRIPTION.rst:1]: Add some background on the modules that would benefit from this technical addons

**Review by chienandalu — APPROVED** [review]

File: website_sale_coupon_reward_line_discriminant/models/sale_order.py (hunk 2 above)
Comment (alvaro-domatix) [reply website_sale_coupon_reward_line_discriminant/models/sale_order.py]: I made all 3 suggested changes. But I would like to ask why this super is better than the other? Or it's just not better, just cleaner?

Comment (pedrobaeza) [reply website_sale_coupon_reward_line_discriminant/models/sale_order.py]: It's a cleaner syntax since Python 3

**Comment by chienandalu** [comment]
@alvaro-domatix Are you still on this?

**Comment by alvaro-domatix** [comment]
Yes, if it is everything fine and I just need to do the test I will do it.

**Comment by alvaro-domatix** [comment]
Hello @chienandalu, I'd like to know if the pre-commit failing has something to do with our code or how to fix it.

**Comment by alvaro-domatix** [comment]
Didn't want to close this PR, can be reopened or do I have to open a new one? @chienandalu @pedrobaeza

**Comment by alvaro-domatix** [comment]
How can we relaunch the checks? @chienandalu @pedrobaeza

**Comment by pedrobaeza** [comment]
do rebase and force push

**Review by chienandalu — APPROVED** [review]

File: website_sale_coupon_reward_line_discriminant/readme/CONTRIBUTORS.rst [hunk 4]
```diff
@@ -0,0 +1,9 @@
+* `Domatix <https://www.domatix.com>`_:
+
+  * Álvaro López
+
+
+* `Tecnativa <https://www.tecnativa.com>`_:
```
Comment (pedrobaeza) [inline website_sale_coupon_reward_line_discriminant/readme/CONTRIBUTORS.rst:6]: No need to include us, as this module is yours.

File: website_sale_coupon_reward_line_discriminant/readme/DESCRIPTION.rst [hunk 5]
```diff
@@ -0,0 +1,2 @@
...
+This module is useful when used in conjunction with others like 'sale_coupon_multi_gift' or 'sale_coupon_multiplier_free_product', allowing you to make offers and the customer to buy more if they want.
```
Comment (pedrobaeza) [inline website_sale_coupon_reward_line_discriminant/readme/DESCRIPTION.rst:2]: Leave an empty line in between for proper recognition of a new paragraph.

**Review by pedrobaeza — CHANGES_REQUESTED** [review]
Very few formal questions. Apart from the comments inline, please name the commit properly:

```
[ADD] sale_coupon_reward_line_driscriminant

Enable the possibility of adding more products of a gifted product as new line
in the order.
```

**Pull request closed without merging**